src/contracts/core/PaymentCoordinator.sol

```diff
@@ -110,88 +106,328 @@ contract PaymentCoordinator is
         for (uint256 i = 0; i < rangePayments.length; i++) {
             RangePayment calldata rangePayment = rangePayments[i];
             bytes32 rangePaymentHash = _payForRange(rangePayment);
-            emit RangePaymentCreated(msg.sender, rangePaymentHash, rangePayment);
+            emit RangePaymentCreated(msg.sender, paymentNonce, rangePaymentHash, rangePayment);
         }
     }
 
     /**
      * @notice similar to `payForRange` except the payment is split amongst *all* stakers
      * rather than just those delegated to operators who are registered to a single avs
+     * @param rangePayments The range payments being created
      */
     function payAllForRange(
         RangePayment[] calldata rangePayments
     ) external onlyWhenNotPaused(PAUSED_PAY_ALL_FOR_RANGE) onlyPayAllForRangeSubmitter {
         for (uint256 i = 0; i < rangePayments.length; i++) {
             RangePayment calldata rangePayment = rangePayments[i];
             bytes32 rangePaymentHash = _payForRange(rangePayment);
-            emit RangePaymentForAllCreated(msg.sender, rangePaymentHash, rangePayment);
+            emit RangePaymentForAllCreated(msg.sender, paymentNonce, rangePaymentHash, rangePayment);
         }
     }
 
+    /**
+     * @notice Claim payments against a given root (read from distributionRoots[claim.rootIndex])
+     * @param claim The PaymentMerkleClaim to be processed.
+     * Contains the root index, earner, payment leaves, and required proofs
+     * @dev only callable by the valid claimer, that is
+     * if claimerFor[claim.earner] is address(0) then only the earner can claim, otherwise only
+     * claimerFor[claim.earner] can claim the payments.
+     */
+    function processClaim(PaymentMerkleClaim calldata claim) external onlyWhenNotPaused(PAUSED_CLAIM_PAYMENTS) {
+        _processClaim(claim);
+    }
+
+    /**
+     * @notice Creates a new distribution root
+     * @param root The merkle root of the distribution
+     * @param paymentCalculationStartTimestamp The start timestamp which payments have been calculated from
+     * @param paymentCalculationEndTimestamp The timestamp until which payments have been calculated
+     * @param activatedAt timestamp at which that the root can be claimed against
+     * @dev Only callable by the paymentUpdater
+     */
+    function submitRoot(
+        bytes32 root,
+        uint64 paymentCalculationStartTimestamp,
+        uint64 paymentCalculationEndTimestamp,
+        uint64 activatedAt
+    ) external onlyPaymentUpdater {
+        uint32 rootIndex = uint32(distributionRoots.length);
+        distributionRoots.push(
+            DistributionRoot({
+                root: root,
+                activatedAt: activatedAt,
```

Comment:
activatedAt should be set to `block.timestamp + activationDelay`

Reply:
Seems a bit unintuitive. So do we also emit block.timestamp + activationDelay? 
The activation delay is only relevant for the paymentUpdater so if we wanted to ensure a later claiming timestamp we could have just submitted a root with a later activatedAt timestamp. Not entirely sure the point of this delay unless you want to have a require like 
```
require(activatedAt - block.timestamp > activationDelay)
```
To ensure a submitted root has at minimum activationDelay number of seconds until the root can be active.

Reply:
the point of enforcing the delay at the contract level is that, if the payment updater posts an incorrect root, then users can withdraw and governance can pause within the delay.

we can do this either by adding the delay when the root is posted or adding the delay in every claim (like we do in DM withdrawals). the former is less gas than the latter.

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/blob/8bcbaf830523cf49949011d507b4750cd221ad67/src/contracts/core/PaymentCoordinator.sol#L151